### PR TITLE
MAINT: Update meson.build files from main branch

### PR DESCRIPTION
**Pull request by charris**
This updates the meson.build files from main in order to get a start on the meson build infrastructure we will need when Python 3.12.0rc1 comes out. In addition to the `meson.build` files in numpy itself, the submodules `numpy/core/src/npysort/x86-simd-sort` and `numpy/core/src/umath/svml` have also been updated to get the meson build changes in them.

The numpy include files have not been updated, as the main change seems to be removal of some of them in line with the changes in 2.0.0. Likewise the ABI and API numbers in `numpy/core/meson.build` have been kept in line with 1.25.x, but the dependencies there should be checked by someone more familiar with the new build infrastructure.

### CODE REVIEW

**Review by rgommers — APPROVED** [review]
Thanks @charris. I pushed the fixes that should hopefully get CI to green. LGTM when CI passed.

This seems good for `1.25.x` indeed. A follow-up PR that's only for `1.26.x` should backport the changes to `pyproject.toml` and CI jobs.
And gh-24153 may be useful to backport in either that same PR (since it also touches CI jobs) or a second follow-up PR.

**Comment by rgommers** [comment]
Not sure what the deal is with the GCC 7 in the `old_gcc` job not liking the SVML update; GCC 7 still passes on `main` so I guess it requires some changes unrelated to the Meson ones.

**Comment by charris** [comment]
There are a bunch of asm errors like:
```
2023-07-16T10:45:40.2237039Z numpy/core/src/umath/svml/linux/avx512/svml_z0_acos_h_la.s: Assembler messages:
2023-07-16T10:45:40.2238551Z numpy/core/src/umath/svml/linux/avx512/svml_z0_acos_h_la.s:61: Error: no such instruction: `vmulph {rn-sae},%zmm0,%zmm0,%zmm2'
```
Not sure what that means here, but maybe a fault in instruction detection? There has been a lot of new SIMD work in the 2.0.0 branch. @rdevulap @sayed Thoughts?

**Comment by charris** [comment]
> I pushed the fixes that should hopefully get CI to green

Thanks Ralf, I was counting on that :)

**Comment by r-devulap** [comment]
Its because its probably using an older version of binutils which doesnt recognize the new AVX512 FP16 ISA. One easy way to get around this is to exclude all SVML file ending with `_h_la.s` (we aren't using those files until we merge PR #21955). main branch setup.py has a way to determine binutils version (see https://github.com/numpy/numpy/commit/12d0df7b7e3cb7c5fb9512d4c9b56be04047b12d) which looks like isn't perfect either (see https://github.com/numpy/numpy/issues/24195#issue-1806689274)

**Comment by charris** [comment]
> easy way to get around this is to exclude all SVML file ending with _h_la.s

That does sound like the easiest way forward. We don't need 1.26 to have all the latest and greatest, we just need to be able to build for Python 3.12 using meson. What is needed to ignore the `_h_la.s ` files, I assume it is somewhere in the configuration checks.

**Comment by r-devulap** [comment]
I'm a little confused, the SVML sources in meson.build do not seem to include the FP16 `*_h_la.s` files: https://github.com/numpy/numpy/blob/eef080a3f2eedc304e0209eed9829b1ddf405da6/numpy/core/meson.build#L866-L910. 

I am unsure where this error is coming from.

**Comment by rgommers** [comment]
> the submodules `numpy/core/src/npysort/x86-simd-sort` and `numpy/core/src/umath/svml` have also been updated to get the meson build changes in them.

If it's not easy to find, you can also just not update the `svml` submodule. Things work on `main` after all, and not that much can have changed since branching 1.25.x

**Comment by r-devulap** [comment]
Ah, sorry. I didnt realize this is updating 1.25.x branch and the old_gcc CI uses the setup.py to build. In that case, you could just port this commit and this should fix the build error: https://github.com/numpy/numpy/commit/12d0df7b7e3cb7c5fb9512d4c9b56be04047b12d

**Comment by charris** [comment]
Hmm, not sure what the problem is with `Cirrus CI / macosx_arm64 CIBW_BUILD:cp39-*`. It is a random failure that I have seen before, might be a Cirrus problem. Rerunning.

**Comment by charris** [comment]
Merged. Thanks @rgommers and @r-devulap .

**Pull request merged**
